### ISSUE TITLE
Real-world Balancer usage question

### ISSUE THREAD

**Issue by shenberg** [facebookresearch/encodec#18]
## ❓ Questions

Hi, first of all, thanks for sharing the code! It's quite well-written.

I've been trying to understand the practical use of the `Balancer` class.

In pseudo-code, what I understood from the documentation is that my use of it in a real training loop should look something like this:

```python
z = encoder(input)
reconstruction = decoder(z)
loss_l1 = F.l1_loss(reconstruction, input)
loss_l2 = F.l2_loss(reconstruction, input)
balancer.backward({'l1' : loss_l1, 'l2' : loss_l2}, reconstruction)
```

So far so good?

Assuming I understood correctly, I still run into an issue once I add a loss term on the latent representation (`z`) in the above. e.g.
```python
z = encoder(input)
reconstruction = decoder(z)
loss_l1 = F.l1_loss(reconstruction, input)
loss_l2 = F.l2_loss(reconstruction, input)
loss_latent = some_regularization_loss(z)
balancer.backward({'l1' : loss_l1, 'l2' : loss_l2, 'z' : some_regularization_loss}, reconstruction)
```

Now, the loss on `z` does not depend on the reconstruction so this won't work. An alternative is to pass `z` to `balancer.backward()` however that comes with the cost of backpropagating through the decoder multiple times. What did you do regarding the quantization loss?

**Comment by shenberg** [comment]
To follow up, I've had some success by using the balancer for all the losses based on the final output of the network and balancing the balancer-balanced loss with internal outputs manually, the old-fashioned multiply-by-constant way 🤷

**Comment by adefossez** [comment]
The balancer is only used for losses that are defined with respect to the output of the model. Other losses contributions must be computed in a separate backward.

**Comment by shenberg** [comment]
Thanks. I see now that it's also mentioned in the paper, so extra thanks for the patience in responding!

**Comment by shenberg** [comment]
I used it in code that isn't public and has been superceded by now. Take a look at [https://github.com/acids-ircam/RAVE](https://github.com/acids-ircam/RAVE) for a good example of usage.
